### ISSUE TITLE
Missing ".target/jniNative.so" file

### ISSUE THREAD

**Issue by neozhang307** [ytchen0323/cloud-scale-bwamem#7]
I run the program with an new argument:
spark-submit --driver-memory 48g --executor-memory 40g --total-executor-cores 8 --master local[20] --driver-java-options "-XX:+PrintFlagsFinal" --conf spark.driver.cores=8 --conf spark.driver.maxResultSize=40g --conf spark.storage.memoryFraction=0.7 --conf spark.eventLog.enabled=true --conf spark.eventLog.dir="eventLogs" --conf spark.akka.threads=8 --conf spark.akka.frameSize=1024 cloud-scale-bwamem-0.2.2-assembly.jar cs-bwamem -bPSW 1  -oChoice 2 -oPath ~/data/HCC1954_100Mreads_tinyPart_1folder.adam  1 ~/data/hg19bwaidx ~/data/hg19

And then the program process for a while,  untill there comes another problem and the program crash again, which is:
java.lang.UnstatisfiedLinkError:  Expecting an absolute path of library:  ./target/iniNative.so

and I cannot find the file in the target directory, where can i find the file?

**Comment by neozhang307** [comment]
I mean didn't the jni parameter optional?

**Comment by PengWeiPRC** [comment]
First, yes, it is optional to use jniNative.so or not, but it is
recommended since it will boost performance considerably.

The command line switch is "-bPSWJNI <1 or 0>". The JNI library is used
defaultedly with "-bPSWJNI 1" and can be turned off by using "-bPSWJNI 0".

Second, I guess that there might be something wrong with the compilation
flow. Take a look at the compilation log, you might find that there is a
"building failure" occurred, which is because the compilation of JNI
libraries needs absolute system path in the pom.xml files.

I have made a modification on the two pom.xml files, one is located at
"cloud-scale-bwamem/src/pom.xml", and the other is at
"cloud-scale-bwamem/src/main/jni_fpga/pom.xml". Please:
1. check the two files out from github
2. modify the "<systemPath>" label in these two files using your own
   absolute path.
3. try to compile it to see if the ".so" files are generated in
   cloud-scale-bwamem/target/

**Comment by neozhang307** [comment]
Thanks, it works now. I can find .so files, including jniNative.so and jniSWExtend.so, the the program process.
But it seems I stuck in the post-process then. After cs-bwamem process, and I got an output, saved in ~/data/small_output. Then I try the merge and sort command, both get a "Unknown command lines arguments error"
the scripts is:
cloud-scale-bwamem-0.2.2-assembly.jar sort /home/cloudera/data/small_data /home/cloudera/data/small_final

**Comment by peterpengwei** [comment]
As far as I know, sorting has not been integrated into CS-BWAMEM. We
generate data with ADAM or SAM format, which can be used as input for UC
Berkeley's ADAM and Dr. Heng Li's BWA, respectively.

**Comment by neozhang307** [comment]
Well, I am not quite understand. 
I mean, after the execution of cs-bwamem, I got a directory. Though I am not familiar with ADAM format, I do know that SAM formate is a file not a directory. 
Besides, you said '...(output) data ... can be used as input .. and Heng Li's BWA', did you meant that cs-bwamem is not a complete alignment workflow, and I can use Heng Li's BWA for further procedures? like merge the directory to a SAM format file?

**Comment by PengWeiPRC** [comment]
Not really. ADAM format is proposed by UC Berkeley as an alternative of the
SAM/BAM format (https://github.com/bigdatagenomics/adam). ADAM format is a
typical HDFS format which uses a directory as a conventional "file". This
is why you see a directory rather than a file.

ADAM has a built-in tool to convert between SAM/BAM format and ADAM format.
It can be used for file conversion.

Moreover, CS-BWAMEM supports both file formats. It can be configured by "
-oChoice". We set it to ADAM by default, as "-oChoice 2", but it can be
modified to "-oChoice 1" so as to generate a SAM file.

**Comment by ytchen0323** [comment]
CS-BWAMEM is a complete flow. It is actually an alternative to BWA(-MEM) for pair-end alignment. As Peng suggested, you can generate SAM file from CS-BWAMEM and then follow the GATK best practice flow. 

In this case, you can use the output file (.sam) generated by CS-BWAMEM and then use SAMtools sort, Picard Markduplicate, GATK (IndelRealignment and BaseRecalibration). We also have the script for this flow just in case you may need that.
